innkeeper/client.go

```diff
@@ -0,0 +1,417 @@
+		&http.Client{Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: o.Insecure}}},
+		make(chan skipper.RawData),
+		make(map[int64]string),
+		make(chan interface{}),
+		make(chan interface{})}
+
+	// start a polling loop
+	go func() {
+		to := time.Duration(0)
+		for {
+			select {
+			case <-time.After(to):
+				to = o.PollTimeout
+				if c.poll() {
+					c.dataChan <- toDocument(c.doc)
+				}
+			case <-c.closer:
+				return
+			}
+		}
+	}()
+
+	return c, nil
+}
+
+func (c *Client) createUrl(path ...string) string {
+	u := *c.baseUrl
+	u.Path = strings.Join(append([]string{""}, path...), "/")
+	return (&u).String()
+}
+
+func parseApiError(r io.Reader) (string, error) {
+	message, err := ioutil.ReadAll(r)
+
+	if err != nil {
+		return "", err
+	}
+
+	ae := apiError{}
+	if err := json.Unmarshal(message, &ae); err != nil {
+		return "", err
+	}
+
+	return ae.ErrorType, nil
+}
+
+func isApiAuthError(error string) bool {
+	switch authErrorType(error) {
+	case authErrorPermission, authErrorAuthentication:
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *Client) authenticate() error {
+	t, err := c.auth.Token()
+	if err != nil {
+		return err
+	}
+
+	c.authToken = t
+	return nil
+}
+
+func (c *Client) requestData(authRetry bool, url string) ([]*routeData, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Authorization", c.authToken)
+	response, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	if response.StatusCode == 401 {
+		apiError, err := parseApiError(response.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if !isApiAuthError(apiError) {
+			return nil, fmt.Errorf("unknown authentication error: %s", apiError)
+		}
+
+		if !authRetry {
+			return nil, errors.New("authentication failed")
+		}
+
+		err = c.authenticate()
+		if err != nil {
+			return nil, err
+		}
+
+		return c.requestData(false, url)
+	}
+
+	if response.StatusCode >= 400 {
+		return nil, fmt.Errorf("failed to receive data: %s", response.Status)
+	}
+
+	routesData, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	result := []*routeData{}
+	err = json.Unmarshal(routesData, &result)
+	return result, err
+}
+
+func getRouteKey(d *routeData) string {
+	return fmt.Sprintf("route%d", d.Id)
+}
+
+func appendFormat(exps []string, format string, args ...interface{}) []string {
+	return append(exps, fmt.Sprintf(format, args...))
+}
+
+func appendFormatHeaders(exps []string, format string, defs []headerData, canonicalize bool) []string {
+	for _, d := range defs {
+		key := d.Name
+		if canonicalize {
+			key = http.CanonicalHeaderKey(key)
+		}
+
+		exps = appendFormat(exps, format, key, d.Value)
+	}
+
+	return exps
+}
+
+func getMatcherExpression(d *routeData) string {
+	m := []string{}
+
+	// there can be only 0 or 1 here, because routes for multiple methods
+	// have been already split
```

Comment:
it is true that this is fishy, but actually i can create a new field for the singular method, and then it is ensured in compile time. good catch.